src/operator/contrib/hawkes_ll-inl.h

```diff
@@ -0,0 +1,493 @@
+  CHECK_EQ(out_attrs->size(), 2U);
+
+  // check ndims
+  CHECK_EQ(in_attrs->at(hawkesll::kMu).ndim(), 2);  // mu (N, K)
+  CHECK_EQ(in_attrs->at(hawkesll::kAlpha).ndim(), 1);  // branching ratio (K,)
+  CHECK_EQ(in_attrs->at(hawkesll::kBeta).ndim(), 1);  // decay exponent (K,)
+  CHECK_EQ(in_attrs->at(hawkesll::kState).ndim(), 2);  // Hawkes states (N, K)
+  CHECK_EQ(in_attrs->at(hawkesll::kIATimes).ndim(), 2);  // i.a. times  (N, T)
+  CHECK_EQ(in_attrs->at(hawkesll::kMarks).ndim(), 2);  // marks (N, T)
+  CHECK_EQ(in_attrs->at(hawkesll::kValidLength).ndim(), 1);  // valid len (N,)
+  CHECK_EQ(in_attrs->at(hawkesll::kMaxTime).ndim(), 1);  // max_time (N,)
+
+  N = in_attrs->at(hawkesll::kIATimes)[0];  // number of samples in batch
+  T = in_attrs->at(hawkesll::kIATimes)[1];  // time length
+  K = in_attrs->at(hawkesll::kMu)[1];  // number of marks
+
+  // check inputs consistent
+  CHECK_EQ(in_attrs->at(hawkesll::kMu)[0], N);
+  CHECK_EQ(in_attrs->at(hawkesll::kMu)[1], K);
+  CHECK_EQ(in_attrs->at(hawkesll::kAlpha)[0], K);
+  CHECK_EQ(in_attrs->at(hawkesll::kBeta)[0], K);
+  CHECK_EQ(in_attrs->at(hawkesll::kState)[0], N);
+  CHECK_EQ(in_attrs->at(hawkesll::kState)[1], K);
+  CHECK_EQ(in_attrs->at(hawkesll::kMarks)[0], N);
+  CHECK_EQ(in_attrs->at(hawkesll::kMarks)[1], T);
+  CHECK_EQ(in_attrs->at(hawkesll::kValidLength)[0], N);
+  CHECK_EQ(in_attrs->at(hawkesll::kMaxTime)[0], N);
+
+  // infer output type
+  SHAPE_ASSIGN_CHECK(*out_attrs, hawkesll::kOutLL, Shape1(N))
+  SHAPE_ASSIGN_CHECK(*out_attrs, hawkesll::kOutStates, Shape2(N, K))
+
+  return out_attrs->at(hawkesll::kOutLL).ndim() != 0U &&
+    out_attrs->at(hawkesll::kOutStates).Size() != 0U;
+}
+
+template<int req>
+struct hawkesll_forward {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i,
+                                  DType* out_loglike,
+                                  DType* out_state,
+                                  const DType* mu,
+                                  const DType* alpha,
+                                  const DType* beta,
+                                  DType* state,
+                                  const DType* lags,
+                                  const int32_t* marks,
+                                  DType* valid_length,
+                                  DType* max_time,
+                                  int K,
+                                  int T,
+                                  DType* temp_register
+                                  ) {
+    int32_t ci;  // current mark
+    DType ll = 0;  // log likelihood
+    DType t = 0;  // current time
+    DType d, lda, comp;  //
+    DType *last_ = &temp_register[i * K];
+    const DType *lag_ = &lags[i * T];
+    const int32_t *mark_ = &marks[i * T];
+    DType *state_ = &out_state[i * K];
+
+    // iterate over points in sequence
+    for (index_t j = 0; j < valid_length[i]; ++j) {
+      ci = mark_[j];
+      t += lag_[j];
+      d = t - last_[ci];
+
```

Comment:
As a general note, it's a good practice to reduce the number of floating point operations by removing duplicated computation. For example, compute `i*K + ci` and `expf(-beta[ci] * d)` once, and store them in a local variable. `expf` is usually expensive

Reply:
Thank you for pointing them out! I think this current version gets rid of those redundancies.